### PR TITLE
Refactor curve.rs

### PR DESCRIPTION
**Pull request by michaelhly**
Making room for the stableswap invariant.

### CODE REVIEW

File: token-swap/program/src/curve/shared.rs [hunk 1]
```diff
@@ -0,0 +1,304 @@
+//! Swap calculations and curve implementations
+
+use solana_program::{
+    program_error::ProgramError,
+    program_pack::{Pack, Sealed},
+};
+
+use super::cp::ConstantProductCurve;
+use super::flat::FlatCurve;
```
Comment (joncinque) [inline token-swap/program/src/curve/shared.rs:9]: Nit: Prefer `use crate::...`

**Review by joncinque — COMMENTED** [review]
Looks good, just a few nits and one suggestion to break up a circular dependency

**Review by joncinque — DISMISSED** [review]
Nice, looks good to me!  Once it passes CI I'll merge it in

**Comment by michaelhly** [comment]
Is it possible to re-run the CI?

**Comment by michaelhly** [comment]
If it compiled it should deploy correctly right?

```
Loading program: ../../target/bpfel-unknown-unknown/release/spl_token_swap.so
Error: Transaction was not confirmed in 10.00 seconds
    at Connection._callee20$ (/home/travis/build/solana-labs/solana-program-library/token-swap/js/node_modules/@solana/web3.js/src/connection.js:2227:13)
    at tryCatch (/home/travis/build/solana-labs/solana-program-library/token-swap/js/node_modules/regenerator-runtime/runtime.js:63:40)
    at Generator.invoke [as _invoke] (/home/travis/build/solana-labs/solana-program-library/token-swap/js/node_modules/regenerator-runtime/runtime.js:293:22)
    at Generator.next (/home/travis/build/solana-labs/solana-program-library/token-swap/js/node_modules/regenerator-runtime/runtime.js:118:21)
    at asyncGeneratorStep (/home/travis/build/solana-labs/solana-program-library/token-swap/js/node_modules/@babel/runtime/helpers/asyncToGenerator.js:3:24)
    at _next (/home/travis/build/solana-labs/solana-program-library/token-swap/js/node_modules/@babel/runtime/helpers/asyncToGenerator.js:25:9)
npm ERR! code 255
npm ERR! path /home/travis/build/solana-labs/solana-program-library/token-swap/js
npm ERR! command failed
npm ERR! command sh -c babel-node --ignore node_modules cli/main.js
npm ERR! A complete log of this run can be found in:
npm ERR!     /home/travis/.npm/_logs/2020-10-30T20_33_18_002Z-debug.log
```

**Comment by CriesofCarrots** [comment]
> Edit: Hmm actually looks like most of the CI runs fail here. 😕

Yes, we've been seeing a lot of flakiness with this in CI lately. It may not be related to your PR

**Comment by michaelhly** [comment]
spl-token broke:
```
  Checking spl-token v2.0.8 (/home/travis/build/solana-labs/solana-program-library/token/program)
error: manual `!RangeInclusive::contains` implementation
    --> token/program/src/instruction.rs:1060:6
     |
1060 |     !(index < MIN_SIGNERS || index > MAX_SIGNERS)
     |      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use: `!(MIN_SIGNERS..=MAX_SIGNERS).contains(&index)`
     |
     = note: `-D clippy::manual-range-contains` implied by `-D warnings`
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#manual_range_contains
error: aborting due to previous error
error: could not compile `spl-token`

```

**Comment by CriesofCarrots** [comment]
Thanks for the heads up, @michaelhly . Looks like a new clippy lint in rust nightly. Our CI needs some love 🤕

**Review by joncinque — APPROVED** [review]
Looks good -- thanks for your patience through the CI updates!

**Pull request merged**